synapse/storage/databases/main/events.py

```diff
@@ -1854,38 +1830,63 @@ def _update_current_state_txn(
-            self.db_pool.simple_upsert_many_txn(
-                txn=txn,
-                table="sliding_sync_membership_snapshots",
-                key_names=("room_id", "user_id"),
-                key_values=[
-                    (room_id, membership_info.user_id)
-                    for membership_info in sliding_sync_table_changes.to_insert_membership_snapshots
-                ],
-                value_names=[
-                    "sender",
-                    "membership_event_id",
-                    "membership",
-                    "event_stream_ordering",
-                    "event_instance_name",
-                ]
-                + list(
-                    sliding_sync_table_changes.membership_snapshot_shared_insert_values.keys()
-                ),
-                value_values=[
+            sliding_sync_snapshot_keys = (
+                sliding_sync_table_changes.membership_snapshot_shared_insert_values.keys()
+            )
+            sliding_sync_snapshot_values = (
+                sliding_sync_table_changes.membership_snapshot_shared_insert_values.values()
+            )
+            # We need to insert/update regardless of whether we have
+            # `sliding_sync_snapshot_keys` because there are other fields in the `ON
+            # CONFLICT` upsert to run (see inherit case (explained in
+            # `_calculate_sliding_sync_table_changes()`) for more context when this
+            # happens).
+            #
+            # XXX: We use a sub-query for `stream_ordering` because it's unreliable to
+            # pre-calculate from `events_and_contexts` at the time when
+            # `_calculate_sliding_sync_table_changes()` is ran. We could be working with
+            # events that were previously persisted as an `outlier` with one
+            # `stream_ordering` but are now being persisted again and de-outliered and
+            # assigned a different `stream_ordering` that won't end up being used. Since
+            # we call `_calculate_sliding_sync_table_changes()` before
+            # `_update_outliers_txn()` which fixes this discrepancy (always use the
+            # `stream_ordering` from the first time it was persisted), we're working
+            # with an unreliable `stream_ordering` value that will possibly be unused
+            # and not make it into the `events` table.
+            txn.execute_batch(
+                f"""
+                INSERT INTO sliding_sync_membership_snapshots
+                    (room_id, user_id, sender, membership_event_id, membership, event_stream_ordering, event_instance_name
+                    {("," + ", ".join(sliding_sync_snapshot_keys)) if sliding_sync_snapshot_keys else ""})
+                VALUES (
+                    ?, ?, ?, ?, ?,
+                    (SELECT stream_ordering FROM events WHERE event_id = ?),
+                    (SELECT instance_name FROM events WHERE event_id = ?)
```

Comment:
We should have fallen back to `"master"` when this isn't filled in

```suggestion
                    (SELECT COALESCE(instance_name, 'master') FROM events WHERE event_id = ?)
```

Fixed up in https://github.com/element-hq/synapse/pull/17636 :fast_forward: